hal/spi_async_api.h

```diff
@@ -0,0 +1,182 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_SPI_API_ASYNC_H
+#define MBED_SPI_API_ASYNC_H
+
+#include "hal/spi_api.h"
+
+#if DEVICE_SPI
+#if DEVICE_SPI_ASYNCH
+
+/** \defgroup hal_spi_async SPI: Asynchronous API.
+ *
+ * This sub API complete the set of functions provided by the SPI API to enable asynchronous access
+ * to the peripheral.
+ *
+ * - Schedule/enqueue a transaction.
+ * - eventually cancel/abort the transaction.
+ * - release the transaction (before or after its completion).
+ *
+ * # Defined behaviour
+ * - `spi_async_transfer_new()` returns NULL if any of `spi_t *obj` or `spi_transfer_args_t *args` is NULL.
+ * - `spi_async_transfer_new()` schedules a transfer using the given parameters.
+ *   All parameters and the embedded references must stay "alive" until completion of the operation.
+ * - `spi_async_transfer_new()` returns a reference counted handle on the scheduled operation.
+ * - `spi_async_transfer_free()` does nothing if passed a NULL pointer.
+ * - `spi_async_transfer_free()` notifies the lowlevel implementation that this reference is no longer
+ *   owned used in the client application (upper/layer code).
+ * - `spi_async_transfer_free()` does **NOT** cancel nor abort a transaction if called before completion.
+ * - `spi_async_transfer_abort()` notifies the lowlevel implementation that the given transaction must be
+ *   cancelled (if not already started) or aborted (if currently running).
+ * - When the operation completes (normally or because of abortion or error) the callback is invoked
+ *   with the provided context and a "reason" describing what triggered the completion.
+ *   This call might be running in an interrupt context and thus all contrainsts applying to ISR
+ *   handler applies to this callback.
+ * - `spi_async_transfer_abort()` does nothing if called on an already completed transaction.
+ * - `spi_async_transfer_abort()` may not wait for the transaction to be cancelled/aborted and returns
+ *   immediatly.
+ * - `spi_free()` cancels and aborts all transactions enqueued for this spi_t.
+ * - the callback passed in spi_async_transfer_new is called only once.
+ *
+ * # Undefined behaviour
+ * - Calling `spi_async_free_handle()` more than one time on a `spi_async_tranfer_t`.
+ * - Calling `spi_async_abort()` after calling `spi_async_free_handle()`.
+ *
+ * # What this API does not cover
+ * The following elements are not covered by this API and are considered implementation details :
+ * - The use of Interrupts and/or DMA for async operations.
+ * - The way `spi_transfer()` is implemented : using specific to this function or using
+ *   `spi_async_transfer()`.
+ * - Wether the SS pin is controlled by hardware or software.
+ * - The way `spi_async_abort()` is implemented : abortion callback invoked in the same thread before
```

Comment:
Should the callback be executed at all in the case of an abort? If higher level code is calling abort in preparation to free the spi object does the current thread need to block waiting for the callback before free?

Reply:
Yes to keep a consistent behaviour. The callback is always invoked. The argument gives a hint on whether the operation was successful or not but the behaviour is always :
1. `spi_async_init` schedule a transfer
2. when the transfer is terminated the callback is invoked.